File: src/omnia/data/cache/SimpleIntCacher.java
package omnia.data.cache;

import static java.util.Objects.requireNonNull;

import java.util.function.IntSupplier;

/**
 * A {@link CachedInt} implementation that uses the given {@link IntSupplier} to provide the value
 * to be cached. This class is not thread-safe.
 *
 * The given {@link IntSupplier#getAsInt()} method is never invoked if the value is cached.
 */
final class SimpleIntCacher implements CachedInt {
  private final IntSupplier supplier;
  private boolean isValid;
  private int value;

  SimpleIntCacher(IntSupplier supplier) {
    this.supplier = requireNonNull(supplier);
  }

  @Override
  public int value() {
    if (!isValid) {
      value = supplier.getAsInt();
      isValid = true;
    }
    return value;
  }

  @Override
  public void invalidate() {
    isValid = false;
  }
}

File: src/omnia/data/cache/SimpleIntCacher.kt
package omnia.data.cache


/**
 * A [CachedInt] implementation that uses the given [IntSupplier] to provide the value
 * to be cached. This class is not thread-safe.
 *
 * The given [IntSupplier.getAsInt] method is never invoked if the value is cached.
 */
internal class SimpleIntCacher(private val supplier: () -> Int) : CachedInt {

  private var isValid = false
  private var value = 0
  override fun value(): Int {
    if (!isValid) {
      value = supplier()
      isValid = true
    }
    return value
  }

  override fun invalidate() {
    isValid = false
  }
}